File: examples/implied_vol.cpp
#include "derivkit/derivkit.hpp"

#include <cmath>
#include <iomanip>
#include <iostream>

int main() {
    using namespace derivkit;

    const VanillaSpec spec{
        .spot = 100.0,
        .strike = 100.0,
        .rate = 0.05,
        .dividend = 0.0,
        .vol = 0.20,
        .time = 1.0,
        .type = OptionType::Call,
    };

    const double true_vol = spec.vol;
    const double px = bs::price(spec);

    VanillaSpec probe = spec;
    probe.vol = 0.50;  // deliberately bad Newton start
    const auto iv = bs::implied_vol(probe, px);

    std::cout << std::setprecision(12);
    std::cout << "Market price (true σ = " << true_vol << "): " << px << "\n";
    std::cout << "Implied vol:  " << iv.vol << "   residual=" << iv.residual
              << "   iters=" << iv.iterations << "   via " << iv.method
              << (iv.converged ? "  converged\n" : "  FAILED\n");

    std::cout << "\nSmile reconstruction from CRR prices (N=401):\n";
    std::cout << std::setw(8) << "K" << std::setw(14) << "tree px" << std::setw(14) << "iv"
              << std::setw(14) << "|iv-σ|"
              << "\n";
    for (double k : {80.0, 90.0, 100.0, 110.0, 120.0}) {
        VanillaSpec s = spec;
        s.strike = k;
        const double tree_px = tree::cox_ross_rubinstein(s, 401).value;
        s.vol = 0.3;
        const auto recovered = bs::implied_vol(s, tree_px);
        std::cout << std::fixed << std::setprecision(2) << std::setw(8) << k << std::setprecision(8)
                  << std::setw(14) << tree_px << std::setw(14) << recovered.vol << std::scientific
                  << std::setprecision(3) << std::setw(14) << std::abs(recovered.vol - true_vol)
                  << "\n";
    }
    return 0;
}

File: src/derivkit/cli/implied_vol.py
from __future__ import annotations

from derivkit.black_scholes import price
from derivkit.implied_vol import implied_vol
from derivkit.trees import cox_ross_rubinstein
from derivkit.types import OptionType, VanillaSpec


def main() -> int:
    spec = VanillaSpec(
        spot=100.0,
        strike=100.0,
        rate=0.05,
        dividend=0.0,
        vol=0.20,
        time=1.0,
        type=OptionType.CALL,
    )
    true_vol = spec.vol
    px = price(spec)
    probe = VanillaSpec(**{**spec.__dict__, "vol": 0.50})
    iv = implied_vol(probe, px)
    print(f"Market price (true σ = {true_vol}): {px}")
    status = "converged" if iv.converged else "FAILED"
    print(
        f"Implied vol:  {iv.vol}   residual={iv.residual}   "
        f"iters={iv.iterations}   via {iv.method}  {status}"
    )
    print("\nSmile reconstruction from CRR prices (N=401):")
    print(f"{'K':>8}{'tree px':>14}{'iv':>14}{'|iv-σ|':>14}")
    for k in (80.0, 90.0, 100.0, 110.0, 120.0):
        s = VanillaSpec(**{**spec.__dict__, "strike": k})
        tree_px = cox_ross_rubinstein(s, 401).value
        s.vol = 0.3
        recovered = implied_vol(s, tree_px)
        print(f"{k:8.2f}{tree_px:14.8f}{recovered.vol:14.8f}{abs(recovered.vol - true_vol):14.3e}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
